Core/LinkCleaner.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  LinkCleaner.swift
+//  DuckDuckGo
+//
+//  Copyright © 2021 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public class LinkCleaner {
+    
+    private var lastAmpUrl: String?
+    
+    public init() {
+        
+    }
+    
+    public func urlIsExtractableAmpLink(_ url: URL,
+                                        config: PrivacyConfiguration) -> String? {
+        let ampFormats = TrackingLinkSettings(fromConfig: config).ampLinkFormats
+        for format in ampFormats {
+            if url.absoluteString.matches(pattern: format) {
+                return format
+            }
+        }
+        
+        return nil
+    }
+    
+    public func isURLExcluded(url: URL, config: PrivacyConfiguration) -> Bool {
+        guard let host = url.host else { return true }
+        
+        if config.isTempUnprotected(domain: host)
+            || config.isUserUnprotected(domain: host)
+            || config.isInExceptionList(domain: host, forFeature: .ampLinks) {
+            return true
+        }
+        
+        return false
+    }
+    
+    public func extractCanonicalFromAmpLink(initiator: URL?, destination url: URL?,
+                                            config: PrivacyConfiguration = PrivacyConfigurationManager.shared.privacyConfig) -> URL? {
```

Comment:
We need to make it resistant to race conditions, as it is possible for user to invoke this by clicking on a link, then tap on another link. There are at least 3 scenarios:

1. Amp Link into Amp Link - so when this is called more than once in succession. We need to cancel "old" extraction and ensure no callbacks are made.
2. Regular link into Amp Link - I'd guess this is not a big problem.
3. Amp Link into Regular link - we would need to cancel the extraction to ensure no callback is made.